### DISCUSSION TITLE
Simple quantifier with array bit vector returns incomplete.

### DISCUSSION THREAD

**Discussion by alexthomasv** [cvc5/cvc5#11446] [Q&A] [answered]
Hello! I am new to CVC5 and what I thought might be a "simple" query to CVC5 resulted in incomplete. I am using the non-pythonic python bindings and the relevant snippet looks like: 

```
      i = self.solver.mkVar(arr.getSort().getArrayIndexSort(), "i")      # Index variable i
      select_term = self.solver.mkTerm(Kind.SELECT, arr, i)
      eq_term = self.solver.mkTerm(Kind.EQUAL, select_term,  
                    self.solver.mkBitVector(arr.getSort().getArrayElementSort().getBitVectorSize(), 0))
      bounded_var = self.solver.mkTerm(Kind.VARIABLE_LIST, i)
      quantified_formula = self.solver.mkTerm(Kind.FORALL, bounded_var, eq_term)
      self.solver.assertFormula(quantified_formula)
      res = self.solver.checkSat()
```
Here are the options I enable:

```
      self.solver = cvc5.Solver()
      self.solver.setLogic("ALL") 
      self.solver.setOption("produce-models", "true")
      self.solver.setOption("produce-unsat-cores", "true")
      self.solver.setOption("minimal-unsat-cores", "true")
      self.solver.setOption("model-cores", "non-implied")
      self.solver.setOption("finite-model-find", "true")
```

This is equivalent to this:

`(assert (forall ((i (_ BitVec 32))) (= (select M i) #b00000000000000000000000000000000)))
`

In Python, i get: `UnknownExplanation.INCOMPLETE`
Where am I going wrong? Any help would be appreciated!

**Comment by ajreynol** [comment, marked as answer]
Sorry for the delay in reply.

The formula you sent is unfortunately cannot be solved by cvc5 using default options, due to its lack of support for constant arrays.

In particular, the model requires cvc5 to interpret `M` as the constant array 0, which the array solver does not have complete support for.

In general, my advice when using quantifiers in cvc5 is summarized here: https://homepage.cs.uiowa.edu/~ajreynol/flowchart-quantifiers.png.  We are planning to make this more available in our documentation.

One note here is that the (experimental) option `--sygus-inference=on` solves your problem, although as a warning this option does not scale well in practice.
